internal/stats/aggregate.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	// Group together metrics based on their groupByFields.
+	groupedMetricBuckets, err := groupTogetherMetrics(metricMapList, query.AggregationQuery)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to group metrics: '%v'.", err)
+	}
+
+	aggregatedResults := make([]map[string]any, 0, len(groupedMetricBuckets))
+
+	// Aggregate on metrics that are grouped together.
+	for _, groupedMetrics := range groupedMetricBuckets {
+		result, err := computeGroupAggregation(groupedMetrics, query)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to compute aggregation: '%v'.", err)
+		}
+
+		if result != nil {
+			aggregatedResults = append(aggregatedResults, result)
+		}
+	}
+
+	return aggregatedResults, nil
+}
+
+func toJSONMetricsMapSlice[T Metric](metrics []T) ([]map[string]any, error) {
+	jsonMetrics, err := util.ToJSON(metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	metricMapList := make([]map[string]any, 0, len(metrics))
+	err = util.JSONFromString(jsonMetrics, &metricMapList)
+	if err != nil {
+		return nil, err
+	}
+
+	return metricMapList, nil
+}
+
+func validateFields(metricType reflect.Type, query AggregationQuery) error {
+	for _, field := range query.GroupByFields {
+		if field == query.OverviewField {
+			return fmt.Errorf("Group by and overview fields must be different.")
+		}
+	}
+
+	if metricType.Kind() == reflect.Ptr {
+		metricType = metricType.Elem()
+	}
+
+	jsonTags := make(map[string]bool)
+	extractJSONTags(metricType, jsonTags)
+
+	for _, field := range query.GroupByFields {
+		if !jsonTags[field] {
+			return fmt.Errorf("Field '%s' is not a valid group-by field.", field)
+		}
+	}
+
+	if !jsonTags[query.OverviewField] {
+		return fmt.Errorf("Field '%s' is not a valid overview field.", query.OverviewField)
+	}
+
+	return nil
+}
+
+func extractJSONTags(reflectType reflect.Type, jsonTags map[string]bool) {
+	for i := range reflectType.NumField() {
+		field := reflectType.Field(i)
+
+		// Recursively get all embedded tags.
+		if field.Anonymous {
+			extractJSONTags(field.Type, jsonTags)
+		}
+
+		jsonTag := util.JSONFieldName(field)
+
+		if jsonTag != "" {
+			jsonTags[jsonTag] = true
+		}
+	}
+}
+
+// Group together metrics based on their groupByFields.
+func groupTogetherMetrics(metrics []map[string]any, query AggregationQuery) (map[string][]map[string]string, error) {
+	groupedMetricBuckets := make(map[string][]map[string]string)
+
+	for _, metric := range metrics {
+		// Skip grouping this metric if it doesn't contain all groupByFields and the overviewField.
+		if !hasAllNeededFields(metric, query) {
+			continue
+		}
+
+		// Turn all values into strings for aggregation.
+		stringMetric := make(map[string]string, len(metric))
+		for field, value := range metric {
+			stringMetric[field] = fmt.Sprintf("%v", value)
+		}
+
+		var groupFieldValues []string
+		for _, field := range query.GroupByFields {
+			groupFieldValues = append(groupFieldValues, fmt.Sprintf("%v", metric[field]))
```

Comment:
I am not a fan of this string conversion, since then you have to convert it back to a number of numeric aggregation.
String conversions can be costly.
Can this just be kept as an any?